### PR TITLE
pre-commit: add `codespell` to check spelling

### PR DESCRIPTION
**Pull request by jbampton**
### Description

This PR adds `codespell` to our pre-commit hooks.

The words in `codespell.txt` are ignored and this file has basically been created by running:

`codespell . | cut -f2 -d' ' | tr A-Z a-z | sort | uniq > codespell.txt`  

from the repo root.  

https://github.com/codespell-project/codespell

`codespell` is one of the leading spell checkers on GitHub.

Going forwards we will need to fix a lot of the misspelled words that are in `codespell.txt`

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [X] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

`pre-commit run codespell --all-files`

Or to run all the hooks:

`pre-commit run --all-files`

### CODE REVIEW

**Comment by DaanHoogland** [comment]
looks good @jbampton . Should we consider the contents of codespell.txt as a list of nouns to review/improve?

**Comment by jbampton** [comment]
Hey @DaanHoogland  yes we need to review `codespell.txt`.    It does have lots of spelling mistakes. 

We can also exclude files/folders from spell checking if needed.

**Comment by DaanHoogland** [comment]
@jbampton , there is a bunch of PRs going on with "polish" in the name. I think this is a bit related to some of that work, And I think we should not exclude any folders. It might be worth to run this without the codespell.txt and analyse how much work it would be to clean it. I think in the end we do want to maintain a short list in that file. I.E. we do want "cloudstack" (not in the file) and "attache" to be allowed for instance. ;)
NOTE: "attache" from french "attaché" is the name of type of a piece of code responsible for managing a resource.
There is bound to be some more in there.
Very interesting addition!

**Review by DaanHoogland — APPROVED** [review]

**Comment by jbampton** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: .github/linters/codespell.txt [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+valide
+valule
+valus
+varibles
+vas
+verfy
+verfying
+verifi
+verifing
+vew
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:636]: why is this in here?

File: .github/linters/codespell.txt [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+vaidate
+valiate
+valide
+valule
+valus
+varibles
+vas
+verfy
+verfying
+verifi
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:634]: I doubt it :)

**Review by DaanHoogland — COMMENTED** [review]
@jbampton any sugestions for followup work on this? I pointed out two obviously wrong words but saw a lot more.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -61,6 +61,14 @@ repos:
...
         files: \.(in|java|md|py|rb|sh|vue|yaml|yml)$
         args: [--markdown-linebreak-ext=md]
+  - repo: https://github.com/codespell-project/codespell
+    rev: v2.2.6
+    hooks:
+      - id: codespell
+        name: run codespell
+        description: Check spelling with codespell
+        args: [--ignore-words=.github/linters/codespell.txt]
+        exclude: ^ui/package\.json$|^ui/package-lock\.json$|^ui/public/js/less\.min\.js$|^ui/public/locales/.*[^n].*\.json$
```
Comment (jbampton) [inline .pre-commit-config.yaml:71]: @DaanHoogland I am now excluding some files from spell checking

**Comment by jbampton** [comment]
@blueorangutan package

File: .github/linters/codespell.txt [hunk 4]
```diff
@@ -52,6 +52,7 @@ authenciation
 authenitcation
 authenitication
 availiability
+avialable
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:55]: shouldn't be in should it?

File: .github/linters/codespell.txt [hunk 5]
```diff
@@ -90,6 +92,7 @@ conectix
 confg
 configruation
 configuable
+conneciton
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:95]: this doesn't seem right either.

File: .github/linters/codespell.txt [hunk 6]
```diff
@@ -405,6 +403,7 @@ reserverd
 reseted
 reseting
 resorce
+responser
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:406]: responder I would say. not sure if this makes sense to allow

File: .github/linters/codespell.txt [hunk 7]
```diff
@@ -440,6 +439,7 @@ shoul
 shoule
 sie
 signle
+simplier
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:442]: should this be simpler?

File: .github/linters/codespell.txt [hunk 8]
```diff
@@ -484,6 +483,7 @@ thorugh
 throught
 ths
 tipically
+transction
```
Comment (DaanHoogland) [inline .github/linters/codespell.txt:486]:
```suggestion
transaction
```

**Review by DaanHoogland — COMMENTED** [review]
@jbampton I like this kind of ruling but making it consistent is going to ask a lot, is it? as said earlier I still see a lot of danglish and franglish..

**Comment by jbampton** [comment]
Hey @DaanHoogland  the list of words in `.github/linters/codespell.txt` will need fixing in other PRs in future.

Basically codespell ignores those words when it runs the spell check. 

So this PR is just adding the codespell test and making sure it passes.

You can see that on the last commit in this PR we are regressing and new misspelled words have been added.

https://github.com/apache/cloudstack/pull/8197/commits/7384e8220cda32f5674e9133527a3e514f802858

So until we get this PR merged misspelled words will continue to be added the codebase.

In future as we clean up and and fix spelling we will recreate the ignored word list hopefully with less misspelled words each time.

**Comment by JoaoJandre** [comment]
If I understood correctly, the `.github/linters/codespell.txt` file has a list of misspelled words that were found within the project and that are going to be ignored by codespell, right? 
If so, should we have a plan for removing and fixing these, so that eventually the file is empty?

**Comment by jbampton** [comment]
Hey @JoaoJandre yes you are correct.

I have been working on spelling fixes for a while now and have a couple of PRs in progress with spell fixes.

**Review by JoaoJandre — APPROVED** [review]
LGTM. I saw some PRs that fixed spellings in the past and always thought that they would eventually have to be redone as new misspellings are introduced constantly. With this change we should eventually be able to fix every misspell on the codebase :)

**Pull request merged**